File: src/api/kanban.ts
import { callCommand } from './client'
import type { RowVersion, Uuid } from './types'

export type PrioridadTarjeta = 'Baja' | 'Normal' | 'Alta' | 'Urgente'
export type TipoPresetTablero = 'trabajos' | 'ordenes'

export interface KanbanTableroDto {
  id: Uuid
  nombre: string
  descripcion: string | null
  color: string | null
  esPreset: boolean
  tipoPreset: TipoPresetTablero | null
  activo: boolean
  rowVersion: RowVersion
}

export interface KanbanColumnaDto {
  id: Uuid
  tableroId: Uuid
  nombre: string
  color: string | null
  orden: number
  limiteWip: number | null
  estadoMapeado: number | null
  rowVersion: RowVersion
}

export interface KanbanEtiquetaDto {
  id: Uuid
  nombre: string
  color: string
  rowVersion: RowVersion
}

export interface KanbanChecklistDto {
  id: Uuid
  tarjetaId: Uuid
  titulo: string
  completada: boolean
  orden: number
  rowVersion: RowVersion
}

export interface KanbanTarjetaDto {
  id: Uuid
  columnaId: Uuid
  titulo: string
  descripcion: string | null
  prioridad: PrioridadTarjeta
  fechaVencimiento: string | null
  orden: number
  trabajoId: Uuid | null
  ordenTrabajoId: Uuid | null
  archivada: boolean
  rowVersion: RowVersion
  etiquetas: KanbanEtiquetaDto[]
  totalChecklist: number
  completadasChecklist: number
}

export interface KanbanTableroDetalleDto {
  tablero: KanbanTableroDto
  columnas: KanbanColumnaDto[]
  tarjetas: KanbanTarjetaDto[]
  etiquetas: KanbanEtiquetaDto[]
}

// --- Inputs ---

export interface CrearTableroInput {
  nombre: string
  descripcion?: string | null
  color?: string | null
}

export interface ActualizarTableroInput {
  nombre: string
  descripcion?: string | null
  color?: string | null
  activo: boolean
  rowVersion: RowVersion
}

export interface CrearColumnaInput {
  tableroId: Uuid
  nombre: string
  color?: string | null
  limiteWip?: number | null
}

export interface ActualizarColumnaInput {
  nombre: string
  color?: string | null
  orden: number
  limiteWip?: number | null
  rowVersion: RowVersion
}

export interface CrearTarjetaInput {
  columnaId: Uuid
  titulo: string
  descripcion?: string | null
  prioridad: PrioridadTarjeta
  fechaVencimiento?: string | null
  trabajoId?: Uuid | null
  ordenTrabajoId?: Uuid | null
  etiquetaIds?: Uuid[] | null
}

export interface ActualizarTarjetaInput {
  titulo: string
  descripcion?: string | null
  prioridad: PrioridadTarjeta
  fechaVencimiento?: string | null
  etiquetaIds?: Uuid[] | null
  rowVersion: RowVersion
}

export interface MoverTarjetaInput {
  tarjetaId: Uuid
  nuevaColumnaId: Uuid
  nuevoOrden: number
  rowVersion: RowVersion
}

export interface CrearEtiquetaInput {
  nombre: string
  color: string
}

export interface ActualizarEtiquetaInput {
  nombre: string
  color: string
  rowVersion: RowVersion
}

export interface CrearChecklistInput {
  tarjetaId: Uuid
  titulo: string
}

export interface ActualizarChecklistInput {
  titulo: string
  completada: boolean
  orden: number
  rowVersion: RowVersion
}

export const kanbanApi = {
  listTableros: () => callCommand<KanbanTableroDto[]>('kanban_list_tableros'),
  getTablero: (id: Uuid) => callCommand<KanbanTableroDetalleDto>('kanban_get_tablero', { id }),
  createTablero: (input: CrearTableroInput) =>
    callCommand<KanbanTableroDto>('kanban_create_tablero', { input }),
  updateTablero: (id: Uuid, input: ActualizarTableroInput) =>
    callCommand<KanbanTableroDto>('kanban_update_tablero', { id, input }),
  deleteTablero: (id: Uuid, rowVersion: RowVersion) =>
    callCommand<void>('kanban_delete_tablero', { id, rowVersion }),

  createColumna: (input: CrearColumnaInput) =>
    callCommand<KanbanColumnaDto>('kanban_create_columna', { input }),
  updateColumna: (id: Uuid, input: ActualizarColumnaInput) =>
    callCommand<KanbanColumnaDto>('kanban_update_columna', { id, input }),
  deleteColumna: (id: Uuid, rowVersion: RowVersion) =>
    callCommand<void>('kanban_delete_columna', { id, rowVersion }),

  createTarjeta: (input: CrearTarjetaInput) =>
    callCommand<KanbanTarjetaDto>('kanban_create_tarjeta', { input }),
  updateTarjeta: (id: Uuid, input: ActualizarTarjetaInput) =>
    callCommand<KanbanTarjetaDto>('kanban_update_tarjeta', { id, input }),
  moverTarjeta: (input: MoverTarjetaInput) =>
    callCommand<void>('kanban_mover_tarjeta', { input }),
  deleteTarjeta: (id: Uuid, rowVersion: RowVersion) =>
    callCommand<void>('kanban_delete_tarjeta', { id, rowVersion }),

  sincronizarPreset: (tableroId: Uuid) =>
    callCommand<void>('kanban_sincronizar_preset', { tableroId }),

  listEtiquetas: () => callCommand<KanbanEtiquetaDto[]>('kanban_list_etiquetas'),
  createEtiqueta: (input: CrearEtiquetaInput) =>
    callCommand<KanbanEtiquetaDto>('kanban_create_etiqueta', { input }),
  updateEtiqueta: (id: Uuid, input: ActualizarEtiquetaInput) =>
    callCommand<KanbanEtiquetaDto>('kanban_update_etiqueta', { id, input }),
  deleteEtiqueta: (id: Uuid, rowVersion: RowVersion) =>
    callCommand<void>('kanban_delete_etiqueta', { id, rowVersion }),

  listChecklist: (tarjetaId: Uuid) =>
    callCommand<KanbanChecklistDto[]>('kanban_list_checklist', { tarjetaId }),
  addChecklistItem: (input: CrearChecklistInput) =>
    callCommand<KanbanChecklistDto>('kanban_add_checklist_item', { input }),
  updateChecklistItem: (id: Uuid, input: ActualizarChecklistInput) =>
    callCommand<KanbanChecklistDto>('kanban_update_checklist_item', { id, input }),
  deleteChecklistItem: (id: Uuid) =>
    callCommand<void>('kanban_delete_checklist_item', { id }),
}

File: src-tauri/src/commands/kanban.rs
//! Tauri commands for the Kanban module.

use certaro_application::dtos::kanban::{
    ActualizarChecklistInput, ActualizarColumnaInput, ActualizarEtiquetaInput,
    ActualizarTableroInput, ActualizarTarjetaInput, CrearChecklistInput, CrearColumnaInput,
    CrearEtiquetaInput, CrearTableroInput, CrearTarjetaInput, KanbanChecklistDto, KanbanColumnaDto,
    KanbanEtiquetaDto, KanbanTableroDetalleDto, KanbanTableroDto, KanbanTarjetaDto,
    MoverTarjetaInput, ReordenarColumnasInput, ReordenarTarjetasInput,
};
use tauri::State;
use uuid::Uuid;

use crate::error::{handle, ApiResult};
use crate::state::AppState;

#[tauri::command]
pub async fn kanban_list_tableros(state: State<'_, AppState>) -> ApiResult<Vec<KanbanTableroDto>> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.list_tableros().await,
        Err(e) => Err(e),
    };
    handle("kanban_list_tableros", outcome)
}

#[tauri::command]
pub async fn kanban_get_tablero(
    state: State<'_, AppState>,
    id: Uuid,
) -> ApiResult<KanbanTableroDetalleDto> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.get_tablero_detalle(id).await,
        Err(e) => Err(e),
    };
    handle("kanban_get_tablero", outcome)
}

#[tauri::command]
pub async fn kanban_create_tablero(
    state: State<'_, AppState>,
    input: CrearTableroInput,
) -> ApiResult<KanbanTableroDto> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.create_tablero(input).await,
        Err(e) => Err(e),
    };
    handle("kanban_create_tablero", outcome)
}

#[tauri::command]
pub async fn kanban_update_tablero(
    state: State<'_, AppState>,
    id: Uuid,
    input: ActualizarTableroInput,
) -> ApiResult<KanbanTableroDto> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.update_tablero(id, input).await,
        Err(e) => Err(e),
    };
    handle("kanban_update_tablero", outcome)
}

#[tauri::command]
pub async fn kanban_delete_tablero(
    state: State<'_, AppState>,
    id: Uuid,
    row_version: String,
) -> ApiResult<()> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.delete_tablero(id, &row_version).await,
        Err(e) => Err(e),
    };
    handle("kanban_delete_tablero", outcome)
}

#[tauri::command]
pub async fn kanban_create_columna(
    state: State<'_, AppState>,
    input: CrearColumnaInput,
) -> ApiResult<KanbanColumnaDto> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.create_columna(input).await,
        Err(e) => Err(e),
    };
    handle("kanban_create_columna", outcome)
}

#[tauri::command]
pub async fn kanban_update_columna(
    state: State<'_, AppState>,
    id: Uuid,
    input: ActualizarColumnaInput,
) -> ApiResult<KanbanColumnaDto> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.update_columna(id, input).await,
        Err(e) => Err(e),
    };
    handle("kanban_update_columna", outcome)
}

#[tauri::command]
pub async fn kanban_delete_columna(
    state: State<'_, AppState>,
    id: Uuid,
    row_version: String,
) -> ApiResult<()> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.delete_columna(id, &row_version).await,
        Err(e) => Err(e),
    };
    handle("kanban_delete_columna", outcome)
}

#[tauri::command]
pub async fn kanban_create_tarjeta(
    state: State<'_, AppState>,
    input: CrearTarjetaInput,
) -> ApiResult<KanbanTarjetaDto> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.create_tarjeta(input).await,
        Err(e) => Err(e),
    };
    handle("kanban_create_tarjeta", outcome)
}

#[tauri::command]
pub async fn kanban_update_tarjeta(
    state: State<'_, AppState>,
    id: Uuid,
    input: ActualizarTarjetaInput,
) -> ApiResult<KanbanTarjetaDto> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.update_tarjeta(id, input).await,
        Err(e) => Err(e),
    };
    handle("kanban_update_tarjeta", outcome)
}

#[tauri::command]
pub async fn kanban_mover_tarjeta(
    state: State<'_, AppState>,
    input: MoverTarjetaInput,
) -> ApiResult<()> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.mover_tarjeta(input).await,
        Err(e) => Err(e),
    };
    handle("kanban_mover_tarjeta", outcome)
}

#[tauri::command]
pub async fn kanban_reordenar_columnas(
    state: State<'_, AppState>,
    input: ReordenarColumnasInput,
) -> ApiResult<()> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.reordenar_columnas(input).await,
        Err(e) => Err(e),
    };
    handle("kanban_reordenar_columnas", outcome)
}

#[tauri::command]
pub async fn kanban_reordenar_tarjetas(
    state: State<'_, AppState>,
    input: ReordenarTarjetasInput,
) -> ApiResult<()> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.reordenar_tarjetas(input).await,
        Err(e) => Err(e),
    };
    handle("kanban_reordenar_tarjetas", outcome)
}

#[tauri::command]
pub async fn kanban_delete_tarjeta(
    state: State<'_, AppState>,
    id: Uuid,
    row_version: String,
) -> ApiResult<()> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.delete_tarjeta(id, &row_version).await,
        Err(e) => Err(e),
    };
    handle("kanban_delete_tarjeta", outcome)
}

#[tauri::command]
pub async fn kanban_sincronizar_preset(
    state: State<'_, AppState>,
    tablero_id: Uuid,
) -> ApiResult<()> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.sincronizar_preset(tablero_id).await,
        Err(e) => Err(e),
    };
    handle("kanban_sincronizar_preset", outcome)
}

#[tauri::command]
pub async fn kanban_list_etiquetas(
    state: State<'_, AppState>,
) -> ApiResult<Vec<KanbanEtiquetaDto>> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.list_etiquetas().await,
        Err(e) => Err(e),
    };
    handle("kanban_list_etiquetas", outcome)
}

#[tauri::command]
pub async fn kanban_create_etiqueta(
    state: State<'_, AppState>,
    input: CrearEtiquetaInput,
) -> ApiResult<KanbanEtiquetaDto> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.create_etiqueta(input).await,
        Err(e) => Err(e),
    };
    handle("kanban_create_etiqueta", outcome)
}

#[tauri::command]
pub async fn kanban_update_etiqueta(
    state: State<'_, AppState>,
    id: Uuid,
    input: ActualizarEtiquetaInput,
) -> ApiResult<KanbanEtiquetaDto> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.update_etiqueta(id, input).await,
        Err(e) => Err(e),
    };
    handle("kanban_update_etiqueta", outcome)
}

#[tauri::command]
pub async fn kanban_delete_etiqueta(
    state: State<'_, AppState>,
    id: Uuid,
    row_version: String,
) -> ApiResult<()> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.delete_etiqueta(id, &row_version).await,
        Err(e) => Err(e),
    };
    handle("kanban_delete_etiqueta", outcome)
}

#[tauri::command]
pub async fn kanban_list_checklist(
    state: State<'_, AppState>,
    tarjeta_id: Uuid,
) -> ApiResult<Vec<KanbanChecklistDto>> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.list_checklist(tarjeta_id).await,
        Err(e) => Err(e),
    };
    handle("kanban_list_checklist", outcome)
}

#[tauri::command]
pub async fn kanban_add_checklist_item(
    state: State<'_, AppState>,
    input: CrearChecklistInput,
) -> ApiResult<KanbanChecklistDto> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.add_checklist_item(input).await,
        Err(e) => Err(e),
    };
    handle("kanban_add_checklist_item", outcome)
}

#[tauri::command]
pub async fn kanban_update_checklist_item(
    state: State<'_, AppState>,
    id: Uuid,
    input: ActualizarChecklistInput,
) -> ApiResult<KanbanChecklistDto> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.update_checklist_item(id, input).await,
        Err(e) => Err(e),
    };
    handle("kanban_update_checklist_item", outcome)
}

#[tauri::command]
pub async fn kanban_delete_checklist_item(
    state: State<'_, AppState>,
    id: Uuid,
) -> ApiResult<()> {
    let outcome = match state.services() {
        Ok(services) => services.kanban.delete_checklist_item(id).await,
        Err(e) => Err(e),
    };
    handle("kanban_delete_checklist_item", outcome)
}
